### PR TITLE
Cut gaussian beam

### PR DESCRIPTION
**Pull request by MaxThevenet**
This PR adds the capability to cut the tail of a gaussian particle beam (typically where the number of macro-particles is low anyway). Sometimes makes it easier for post-processing. It also includes a test and documentation.

### CODE REVIEW

File: Source/Particles/PhysicalParticleContainer.cpp [hunk 1]
```diff
@@ -269,7 +270,10 @@ PhysicalParticleContainer::AddGaussianBeam(Real x_m, Real y_m, Real z_m,
             Real y = 0.;
             Real z = distz(mt);
 #endif
-            if (plasma_injector->insideBounds(x, y, z)) {
+            if (plasma_injector->insideBounds(x, y, z)  &&
+                std::abs( x - x_m ) < x_cut * x_rms     &&
```
Comment (LDAmorim) [inline Source/Particles/PhysicalParticleContainer.cpp]: Thanks for this cool PR @MaxThevenet !
It is very nice to read.
But I am still a bit ignorant so I just don't understand very well the default case, i.e. when the input file doesn't define the `x_cut` parameter.
```cpp
std::numeric_limits<amrex::Real>::max();
```
Does this contain the information on `x_rms`?

This is just a question for curiosity, but would it be also useful to define a cut not via a fraction of beam length, but at a fixed position?

Comment (MaxThevenet) [reply Source/Particles/PhysicalParticleContainer.cpp]: I think with absolute or relative to RMS is fine, as long as it is explicitly mentioned in the doc. I find relative more practical, so I implemented this one.
By default, the beam is cut for something like `std::abs( x - x_m ) < 1.e308 * x_rms `, i.e., it is not cut at all, so if the user doesn't specify `x_cut` the beam isn't cut.

Comment (LDAmorim) [reply Source/Particles/PhysicalParticleContainer.cpp]: Relative is practical yes.
Oh I see. I was thinking of cutting in regards to the box size, but that would me much more confusing.
The way you did it makes a lot of sense now. Thank you for explaining @MaxThevenet !

File: Docs/source/running_cpp/parameters.rst [hunk 2]
```diff
@@ -276,7 +276,7 @@ Particle initialization
       ``<species_name>.x/y/z_m`` (average position in `x/y/z`),
       ``<species_name>.x/y/z_rms`` (standard deviation in `x/y/z`),
       ``<species_name>.x/y/z_rms`` (standard deviation in `x/y/z`),
-      ``<species_name>.x/y/z_cut`` (optional, particles with ``abs(x-x_m) > x_cut*x_rms`` are not injected, same for y and z),
+      ``<species_name>.x/y/z_cut`` (optional, particles with ``abs(x-x_m) > x_cut*x_rms`` are not injected, same for y and z. Note that cutting the distribution is likely to result in a lower total charge),
```
Comment (LDAmorim) [inline Docs/source/running_cpp/parameters.rst]: This is a very good point @MaxThevenet !
I don't know how to explain it. But I would make sure users still initialize the beam with the correct `q_tot`.
Maybe something like this:
```suggestion
      ``<species_name>.x/y/z_cut`` (optional, particles with ``abs(x-x_m) > x_cut*x_rms`` are not injected, same for y and z. Note that cutting the distribution is likely to result in a lower total charge of macro-particles in your output diagnostics. But the `<species_name>.q_tot` parameter should still be the total charge of the particles (without cut).),
```

File: Source/Particles/PhysicalParticleContainer.H [hunk 3]
```diff
@@ -201,6 +201,7 @@ public:
 
     void AddGaussianBeam(amrex::Real x_m, amrex::Real y_m, amrex::Real z_m,
```
Comment (ax3l) [inline Source/Particles/PhysicalParticleContainer.H]: `const`-ify all of those? :)

File: Source/Particles/PhysicalParticleContainer.cpp (hunk 1 above)
Comment (ax3l) [reply Source/Particles/PhysicalParticleContainer.cpp]: oh, now I see why you asked that `Inf` question earlier today :)

Comment (ax3l) [reply Source/Particles/PhysicalParticleContainer.cpp]: above: `Real weight = q_tot/npart/charge` is better written: `Real const weight = q_tot/(npart*charge);` (and similarly a few lines lower)

all the temporary `x`, `y`, `z` vars can also be `const`-ified.

**Comment by MaxThevenet** [comment]
@ax3l thanks I added const. Once this PR is merged, I'll do a bit of work in PhysicalParticleContainer.H/cpp for const and formatting.

File: Examples/Tests/initial_distribution/analysis_distribution.py [hunk 4]
```diff
@@ -96,17 +96,34 @@
...
+f_z = npart*db * np.exp(-0.5*(bin_value/x_rms)**2)/(x_rms*np.sqrt(2.0*scc.pi))
+f_z[ np.absolute(bin_value) > z_cut * x_rms ] = 0.
+f_peak = np.amax(f_z)
 
 # compute error
-f4_error = np.sum(np.abs(f-h4x)+np.abs(f-h4y)+np.abs(f-h4z))/bin_value.size / f_peak
+f4_error = np.sum(np.abs(f_xy-h4x)+np.abs(f_xy-h4y)+np.abs(f_z-h4z))/bin_value.size / f_peak
+
+do_plot = False
+if do_plot:
```
Comment (ax3l) [inline Examples/Tests/initial_distribution/analysis_distribution.py:115]: this is unreachable code by default, maybe add a cmd line arg?

Comment (MaxThevenet) [reply Examples/Tests/initial_distribution/analysis_distribution.py:115]: I agree this piece of code adds a warning, but I am not sure adding lines to make it a command-line argument and document it helps the code, if it's just to remove a warning... If you want to get rid of the warning, I am fine with just deleting the plotting option.

Comment (ax3l) [reply Examples/Tests/initial_distribution/analysis_distribution.py:115]: Ha, you don't get out that easily ;-)
See:
```suggestion
def usage():
    print("usage:")
    print("  {} [-p,--plot]".format(sys.argv[0]))

import sys
import getopt        
try:
    opts, args = getopt.getopt(sys.argv[1:], "hp",
        ["help", "plot"])
except getopt.GetoptError as err:
    print(err)
    usage()
    sys.exit(2)

for opt, arg in opts:
    if opt in ["-h", "--help"]:
        usage()
        sys.exit()
    elif opt in ["-p", "--plot"]:
        do_plot = True

if do_plot:
```

Comment (ax3l) [reply Examples/Tests/initial_distribution/analysis_distribution.py:115]: :'(

**Pull request merged**